kernel/src/schema/diff.rs

```diff
@@ -0,0 +1,1739 @@
+
+        assert_eq!(diff.updated_fields.len(), 1);
+        assert_eq!(
+            diff.updated_fields[0].change_type,
+            FieldChangeType::ContainerNullabilityLoosened
+        );
+        assert!(!diff.has_breaking_changes());
+    }
+
+    #[test]
+    fn test_map_nullability_tightened() {
+        let current = StructType::new_unchecked([create_field_with_id(
+            "lookup",
+            DataType::Map(Box::new(MapType::new(
+                DataType::STRING,
+                DataType::INTEGER,
+                true, // Nullable values
+            ))),
+            false,
+            1,
+        )]);
+
+        let new = StructType::new_unchecked([create_field_with_id(
+            "lookup",
+            DataType::Map(Box::new(MapType::new(
+                DataType::STRING,
+                DataType::INTEGER,
+                false, // Non-nullable now
+            ))),
+            false,
+            1,
+        )]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+
+        assert_eq!(diff.updated_fields.len(), 1);
+        assert_eq!(
+            diff.updated_fields[0].change_type,
+            FieldChangeType::ContainerNullabilityTightened
+        );
+        assert!(diff.has_breaking_changes());
+    }
+
+    #[test]
+    fn test_map_with_struct_key() {
+        // Test that maps with struct keys can be diffed
+        let current = StructType::new_unchecked([create_field_with_id(
+            "lookup",
+            DataType::Map(Box::new(MapType::new(
+                DataType::try_struct_type([create_field_with_id(
+                    "id",
+                    DataType::INTEGER,
+                    false,
+                    2,
+                )])
+                .unwrap(),
+                DataType::STRING,
+                true,
+            ))),
+            false,
+            1,
+        )]);
+
+        let new = StructType::new_unchecked([create_field_with_id(
+            "lookup",
+            DataType::Map(Box::new(MapType::new(
+                DataType::try_struct_type([create_field_with_id(
+                    "identifier", // Renamed key field
+                    DataType::INTEGER,
+                    false,
+                    2,
+                )])
+                .unwrap(),
+                DataType::STRING,
+                true,
+            ))),
+            false,
+            1,
+        )]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+
+        // Should see the nested key field renamed
+        assert_eq!(diff.updated_fields.len(), 1);
+        assert_eq!(
+            diff.updated_fields[0].path,
+            ColumnName::new(["lookup", "key", "identifier"])
+        );
+        assert_eq!(diff.updated_fields[0].change_type, FieldChangeType::Renamed);
+    }
```

Comment:
add some cursed cases like:

`map<array<struct<a int>>, map<struct<b int>, array<int>>>`

These are the most frightening ones lol